Convert mail stack

m1cr0man opened the pull request:
Closes #1 

It's.Finally.Done.

Adds a Postfix, Dovecot, Rspamd and Mailman configuration. All aliases have also been migrated. I guess the most useful thing I can do here is describe some features.

- All users must be authenticated to send and receive mail. Port 25 is only useful for users connecting from domains other than ${tld}
- No mail is received unauthenticated for ${tld}, preventing outside sources spoofing emails to us, from our domain.
- Dovecot2 sieve scripts are configured to teach Rspamd Spam and Ham based on users moving mails to and from their Junk/Spam mailbox.
- Connections from local networks must be authenticated. There's a few reasons for this, 1 is that we do not inherently trust the users inside our network, 2 is it prevent spoofing internally, 3 is it makes it much easier to trace the sources of emails from the postfix logs.
- Mailman has its own LDAP user now (mailmgr) which is the only account allowed to spoof mail.
- Mail is now written to /var/mail/%domain/%user in MDBox format. This reduces the number of files tenfold, meaning performance will be better (less random IO on very small files). Also, the mail attachment directory is common to all users allowing dovecot to hash and deduplicate attachments on receive. For my own inbox alone, 25% of my mail was attachments, most of which were shared.
- /var/mail has come complex permissions to support dsync migrations from Maildir. See [this commit's description](https://github.com/redbrick/nix-configs/commit/66efb1e28b57ce4110e4ad69f097a23350403946)
- Rspamd does not obey HTTP proxy configuration when downloading blacklists, meaning the mail host must be able to reach the internet
- Due to the complexity of configuring Mailman + Hyperkitty, and the current webmail setup, mail + web (Apache) must be on the same host.

m1cr0man committed:
Add mailman

Build mod_Wsgi with python3

Add local mailman vhost, add ldap + db config

Add working hook to populate ldap user

m1cr0man committed:
Change m1vm acme server

m1cr0man committed:
Remove email auth, add superuser group

m1cr0man committed:
Disable mailman signup, prevent sender spoofing

Add smtpd_sender_login_names

Use mapFiles for sender blacklist

Add whitelist for smtpd_sender_restrictions

Allows specific users to bypass the rest of the rules

Don't allow relaying internally from unauthed users

m1cr0man committed:
Separate + configure rspamd and redis

Redis over unix only

Make milter port a variable

m1cr0man committed:
Update smtpd_sender_restrictions to prevent spoofing

m1cr0man committed:
Enable rspamd spf checks

Add more rspamd modules, remove permit_mynetworks

Blacklist local addrs from connecting to postfix unauthed

m1cr0man committed:
Fix mailman auth crash on missing surnames

m1cr0man committed:
Relax postfix connection restrictions for internal hosts

Fix client restrictions blocking internal origins

m1cr0man committed:
Merge changes from branch mail_lastgood

Fix typo

Change skipVhosts to skipCustomVhosts

The goal of this parameter from the start was to allow safe
configuration of a dev host with certs without pulling in vhosts
that are only present on the production host - AKA any non-RB
domains. This commit renames the option and restructures the vhosts

m1cr0man committed:
Fix list.tld vhost and postfix sender map

m1cr0man committed:
Port mail aliases from exim

m1cr0man committed:
Add dovecot sieve scripts

Use dovecot2 as the mail user

This means the mail user and group IDs are fixed across all nix
hosts

Fix compile issues

Add pigeonhole mtime patch

m1cr0man committed:
Prevent LDAP uid override for mail ownership

We don't actually want to use user uid/gid for mail in /var/mail.
It should all be owned by the mail user

Add NFS mount for /var/mail

Add iterate_ ldap settings for mass dsync

Change m1vm mail mount to mailtest

Fix dovecot losing authed user domain name

m1cr0man committed:
Do use LDAP uid as mail owner

In order to perform a migration of maildirs using dsync, it must
setuid as the target user for each migration, read mail from their
~/Maildir and then write it to /var/mail. This isn't possible if
every user's UID is vmail according to Dovecot. Instead, we must
set 3770 permissions on /var/mail and vmail:vmail ownership,
so that dovecot can write mail and create folders with the correct
perms (in particular /var/mail/%d and /var/mail/attachments) and
users can still open their mail, whilst also supporting dsync
migrations.

m1cr0man committed:
Add mail to hardcase

Ensure rb aliases are loaded

butlerx commented on nixops.nix:
```diff
@@ -0,0 +1,5 @@
+{
+  icarus = { config, pkgs, ... }: {
```
just the one box?

m1cr0man replied:
Uh, actually I didn't mean to commit this yet. Whoops! But yes, for now. Just testing it.

butlerx commented on services/dovecot/default.nix:
```diff
@@ -1,24 +1,58 @@
+  # Create the user + group that will own /var/mail
+  # Uid not actually used. If the year is 2021 and this config is in production
+  # you can remove uid=uid from user_attrs in auth.nix and simplify the perms
+  # in /var/mail. Read the blog post for more info
```
blog post?

m1cr0man replied:
Soon :tm:

butlerx commented on services/httpd/default.nix:
```diff
@@ -120,7 +119,6 @@ in {
       </Directory>
 
       AddHandler cgi-script .cgi
-      AddHandler cgi-script .py
```
why?

m1cr0man replied:
It doesn't do anything, afaik? I'll test it.
From what I know the python interpreter isn't and never was wired up properly to Apache.

butlerx replied:
grand just making sure

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+#
+  "founders" = "drjolt, wibble, sandman, fergus, swipe, hyper";
+
+# Committee is now a mailing list (handled by mailman)
```
remove the comment and put in docs instead

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+# Web related #
+#-------------#
+
+# webmaster is now a mailman list
```
remove the now

m1cr0man replied:
Ach missed this. Will add it to my next commit, and the other one.

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+  "help-request" = "helpdesk";
+  "helpdesk-request" = "helpdesk";
+  "faildesk" = "helpdesk";
+  "thosecunts" = "helpdesk";
```
some of these aliases feel less then great to put in git

m1cr0man replied:
I did a quick search for profanities, missed that one..

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+  "spamtastic" = "bubble";
+
+# an anonymous news gateway for redbrick.sex
+#redbrick.sex:           "|/local/bin/gateway"
```
just dump this?

m1cr0man replied:
do we still have the domain? :P

butlerx replied:
no idea

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+  "su-webgroup" = "phil, arioch, p, esoteric";
+
+# The Sensei Go Club
+  "senseigoclub" = "pooka, belial, plop+go";
```
i feel these could be lists

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+#
+# NOTES
+# =====
+# - Run 'make exim_aliases' after editing. If no errors found then check it back into
```
this still true?

m1cr0man replied:
Hah, no. I guess I should clean up the comments in there.

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,589 @@
+#   RCS which also implies -> this file is under RCS!
+#
+
+  "redbrick-greendaybugtraq" = "\"|/srv/admin/scripts/m2n/mail2nntp.pl lists.bugtraq\"";
```
@gruunday this you?

butlerx replied:
also is this monstrosity needed?

m1cr0man replied:
Remember that time you deleted an entire service he wrote?

So no, I guess not :P

butlerx replied:
ive no memory of deleting a service

butlerx commented on services/httpd/vhosts.nix:
```diff
@@ -213,7 +214,10 @@ in (userVhosts // {
   "graphs.${tld}" = vhostProxy "http://localhost:3001";
   "dcufm.${tld}" = vhostProxy "http://136.206.15.74";
   "jakarta.${tld}" = vhostProxy "http://136.206.15.59:8080";
+<<<<<<< HEAD
```
ggit rebase gone wrong

butlerx commented on services/postfix/exim2nix.py:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python3
+# Python 3.6+
+# Converts exim aliases text file to Nix
+# Usage: cat exim_aliases.txt | ./exim2nix.py > aliases.nix
```
Why can't we just convert this once and maintain it in nix

m1cr0man replied:
That's exactly what we've done. I've kept this here because I keep all my scripts and notes on mappings for the off-chance that someone needs to do it again (read: me)

butlerx replied:
i just feel like keeping the txt file and nix file could be confusing for future admins

m1cr0man replied:
I'm only keeping the python script.

butlerx commented on services/postfix/mailman.nix:
```diff
@@ -0,0 +1,158 @@
+# Manual steps post-deploy:
```
i assume one of if the case ensure it's in the docs

m1cr0man replied:
Will be doc'd, along with the rest of the setup

m1cr0man committed:
Fix postfix rejecting mail to aliases

Attempt support sending under alias names

Tidy up aliases

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -185,61 +173,17 @@
-  "redbrick-debian-security" = "\"|/srv/admin/scripts/m2n/mail2nntp.pl lists.debian-security\"";
-  "redbrick-apache" = "\"|/srv/admin/scripts/m2n/mail2nntp.pl lists.apache\"";
-  "redbrick-exim" = "\"|/srv/admin/scripts/m2n/mail2nntp.pl lists.exim\"";
-  "redbrick-php" = "\"ryaner\"";
```
lol he'll be so sad he wont get those emails he doesnt get

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,500 @@
+#
+# NOTES
+# =====
+# - For committee position aliases during the changeover period, list the
```
all the positions should me mailing lists not aliases to avoid this

m1cr0man replied:
I disagree. Positions can be versioned here. Mailing lists come with a lot of overhead, and I don't trust mailman's archive security very much.

butlerx replied:
Fine then plz group all positions emails together under this comment

butlerx commented on services/postfix/aliases.nix:
```diff
@@ -0,0 +1,500 @@
+  "su-webgroup" = "phil, arioch, p, esoteric";
+  "senseigoclub" = "pooka, belial, plop+go";
+  "committee" = "committee@lists.${tld}";
+  "rb-admins" = "rb-admins@lists.${tld}";
```
could we group these with the above that alias to them for admins, and committee buisness

ylmcc reviewed (approved):
Looks good to me

m1cr0man committed:
Stop rspamd stripping redbrick from domain

m1cr0man committed:
Fix bugs found post-deploy

Fix mailman archiving woes

Use unix sockets for sasl + lmtp

Originally, I had thought for a HA setup
that we would need tcp ports for lmtp
and sasl. Our unit of failure is a host,
not a service so this is redundant and
actually quite insecure. Using unix
sockets is easier to configure and more
reliable

Small fixes for mail

- Redirect mail -> webmail
- Make mailman map ldap groups to django
- Fix setuid issues in dovecot imap

butlerx commented:
this was deployed can we get this merged to ensure maintaining the code isnt difficult

m1cr0man commented:
I need to rebase + squash this, which I don't have time to do until the weekend, then it can be merged.

m1cr0man commented:
Right squashed from 80 to 19 commits. Man, that's not many for 6/7 months of work :upside_down_face: 

Ready to merge
